retrofit2/src/main/java/com/linecorp/armeria/client/retrofit2/ArmeriaRetrofitBuilder.java

```diff
@@ -41,127 +39,118 @@
+     *                    // Enforce HTTP/1.
+     *                    final SessionProtocol actualProtocol =
+     *                            protocol.isTls() ? SessionProtocol.H1 : SessionProtocol.H1C;
+     *
+     *                    final EndpointGroup actualEndpointGroup;
+     *                    if ("group-bar".equals(url.host())) {
+     *                        // Client-side load-balancing:
+     *                        // - Make the request go to 'node-1.bar.com' or 'node-2.bar.com'
+     *                        //   if the target host is 'group-bar'.
+     *                        actualEndpointGroup = groupBar;
+     *                    } else {
+     *                        // Use the given host and port otherwise.
+     *                        actualEndpointGroup = Endpoint.of(url.host(), url.port());
+     *                    }
      *
-     * @param configurator a {@link BiFunction} whose first argument is the the URI of the server endpoint and
-     *                     whose second argument is the {@link ClientOptionsBuilder} with default options of
-     *                     the new derived client
+     *                    return WebClient.builder(actualProtocol, actualEndpointGroup)
+     *                                    // Derive most settings from 'defaultWebClient'.
+     *                                    .factory(defaultWebClient.factory())
+     *                                    .options(defaultWebClient.options())
+     *                                    // Set a custom header.
+     *                                    .setHttpHeader(HttpHeaderNames.AUTHORIZATION,
+     *                                                   "bearer my-access-token")
+     *                                    // Override the timeout.
+     *                                    .responseTimeout(Duration.ofSeconds(30))
+     *                                    .build();
+     *                })
+     *                .build();
+     * }</pre></p>
+     *
+     * <p>Note that the specified {@link BiFunction} is not used for sending requests to the base URL's
+     * authority. The default {@link WebClient} specified with {@link ArmeriaRetrofit#of(WebClient)} or
+     * {@link ArmeriaRetrofit#builder(WebClient)} will be used instead for such requests:
+     * <pre>{@code
+     * // No need to use 'nonBaseClientFactory()' method.
+     * ArmeriaRetrofit.of(WebClient.builder("http://example.com/")
+     *                             .setHttpHeader(HttpHeaderNames.AUTHORIZATION,
+     *                                            "bearer my-access-token")
+     *                             .build());
+     * }</pre></p>
      */
-    public ArmeriaRetrofitBuilder withClientOptions(
-            BiFunction<String, ? super ClientOptionsBuilder, ClientOptionsBuilder> configurator) {
-        this.configurator = requireNonNull(configurator, "configurator");
+    public ArmeriaRetrofitBuilder nonBaseClientFactory(
```

Comment:
Since normal usage doesn't need to worry about this, I guess it's ok. But doesn't it seem like too advanced functionality? The asymmetry vs `WebClient` (the one I pointed to above) seems unnatural too - if we really want this functionality, we could move it to a `ClientOption` so it applies to both?

And while it doesn't actually solve the asymmetry problem by itself, I'm wondering whether we could just let the `EndpointSelector` layer take care of it, it seems simpler to reason about and don't even need to worry about a client cache

```java
class HostAwareEndpointGroup implements EndpointGroup {
  public Endpoint select(ClientRequestContext ctx) {
    switch (ctx.request().authority()) {
      case "group-bar":
        return groupBarEndpointGroup.select(ctx);
      ...
```

Reply:
I think `ENDPOINT_REMAPPER` options handles this better. What do you think?